base_layer/wallet/src/connectivity_service/service.rs

```diff
@@ -185,6 +193,9 @@ impl WalletConnectivityService {
                         target: LOG_TARGET,
                         "Base node connection failed: {}. Reconnecting...", e
                     );
+                    if let Some(node_id) = self.current_base_node() {
+                        self.disconnect_base_node(node_id).await;
```

Comment:
btw disconnecting a base can happens automatically after a period of time (unless it happens to be a neighbour) if no one is holding the PeerConnection handle. Not sure if there was a particular reason to disconnect, but you can probably let comms handle that for you

Reply:
Understood. As commented in the PR description, the reason for this is to not have to restart the wallet if for some reason we have a bad connection to start with. The `disconnect_base_node` statements have been selectively inserted only when a bad connection is detected from the client's side. In all the tests I have done for the console wallet, disconnecting a bad base node connection definitely improves communication all around, even in the middle of a stress test, as all dependent protocols are self rectifying.

The log extract below shows such a bad connection where eventually the console wallet had to be restarted after 1.5 hours - for this, the connected base node stayed online and was not restarted and communicated perfectly with other nodes.

```rust
2021-11-16 08:56:02.799226800 [wallet::connectivity] DEBUG Attempting to connect to base node peer ece4c616a156c1e4fe34d94d8c...
2021-11-16 08:57:22.523497500 [wallet::connectivity] DEBUG Successfully established peer connection to base node ece4c616a156c1e4fe34d94d8c
2021-11-16 08:57:36.891108100 [wallet::connectivity] DEBUG Successfully established RPC connection ece4c616a156c1e4fe34d94d8c
2021-11-16 08:57:36.891124100 [wallet::connectivity] DEBUG Wallet is ONLINE and connected to base node ece4c616a156c1e4fe34d94d8c
2021-11-16 09:00:13.084042000 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:01:43.096424800 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:03:13.101571800 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:04:43.111513700 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:06:13.120089200 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:07:43.124323400 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:09:13.129675900 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:10:43.133135700 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:12:13.144324100 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:13:43.154570400 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:15:13.169172200 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:16:43.181434500 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:18:13.182299500 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:21:51.244670000 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:23:21.250904200 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:24:51.263524700 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:26:21.268379800 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:27:51.279295000 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:29:21.283854200 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:30:51.299524200 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:32:21.309984300 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:33:51.320823900 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:35:21.325223200 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:36:51.330991400 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:38:21.347124100 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:39:51.359091600 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:41:21.368998600 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:42:51.381214200 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:44:21.385566300 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:45:51.386209300 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:47:21.387524000 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:48:51.397354900 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:50:21.411220200 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:51:51.414259600 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:53:21.420962400 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:54:51.434485500 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:56:21.449505500 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:57:51.456611100 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 09:59:21.464595700 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 10:00:51.473677500 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 10:02:21.477033700 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 10:03:51.478407900 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 10:05:21.488478100 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 10:06:51.490239500 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 10:08:21.500649700 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 10:09:51.512070400 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 10:11:21.514823900 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 10:12:51.521282700 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 10:14:21.528710300 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 10:15:51.543031900 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 10:17:21.554784400 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 10:18:51.559731900 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 10:20:21.565718000 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 10:21:51.574891500 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 10:23:21.587307500 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 10:24:51.603359900 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...
2021-11-16 10:26:21.609144200 [wallet::connectivity] WARN  Base node connection failed: Failed to create client connection: Handshake error: RPC handshake timed out. Reconnecting...

2021-11-16 11:26:56.432504900 [wallet::connectivity] DEBUG Wallet connectivity service has started.
2021-11-16 11:26:56.432718400 [wallet::connectivity] WARN  1 requests are waiting for base node to be set
2021-11-16 11:26:56.457973900 [wallet::connectivity] DEBUG Attempting to connect to base node peer ece4c616a156c1e4fe34d94d8c...
2021-11-16 11:27:00.660710200 [wallet::connectivity] DEBUG Successfully established peer connection to base node ece4c616a156c1e4fe34d94d8c
2021-11-16 11:27:01.872023300 [wallet::connectivity] DEBUG Successfully established RPC connection ece4c616a156c1e4fe34d94d8c
2021-11-16 11:27:01.872042000 [wallet::connectivity] DEBUG Wallet is ONLINE and connected to base node ece4c616a156c1e4fe34d94d8c
```

Reply:
Ok understood, give a good clue, maybe something we need to do to keep the connection alive - thought occurs that this happens in the base node with ping/pong but not on the console wallet which may be the cause.